aioxmpp/ibb/service.py

```diff
@@ -0,0 +1,507 @@
+                    return
+
+            if len(self._write_buffer) < self._output_buffer_limit_low:
+                if self._protocol is not None:
+                    self._protocol.resume_writing()
+
+            if not self._write_buffer:
+                if self._closing:
+                    break
+                self._can_write.clear()
+
+        close = ibb_xso.Close()
+        close.sid = self._sid
+        stanza = aioxmpp.IQ(
+            aioxmpp.IQType.SET,
+            to=self._peer_jid,
+            payload=close,
+        )
+        e = None
+        try:
+            yield from self._service.client.send(
+                stanza
+            )
+        except errors.XMPPCancelError as e:
+            pass
+        finally:
+            self._connection_closed(e, cancel=False)
+
+    def write(self, data):
+        """
+        Send `data` over the IBB. If `data` is larger than the block size
+        is is chunked and sent in chunks.
+
+        Chunks from one call of :method:`send` will always be sent in
+        series.
+        """
+
+        if self._closing or self._closed:
+            return
+
+        self._write_buffer += data
+
+        if len(self._write_buffer) >= self._output_buffer_limit_high:
+            if self._protocol is not None:
+                self._protocol.pause_writing()
+
+        if self._write_buffer:
+            self._can_write.set()
+
+    def _connection_closed(self, e=None, *, cancel=True):
+        if cancel:
+            self._write_task.cancel()
+        self._closed = True
+        self._service._remove_session(self._peer_jid, self._sid)
+        if self._protocol is not None:
+            self._protocol.connection_lost(e)
+
+    def close(self):
+        """
+        Close the session.
+        """
+        if self._closing:
+            return
+
+        self._closing = True
+        # make sure the writer wakes up
+        self._can_write.set()
+
+    def abort(self):
+        """
+        Abort the session.
+        """
+        if self.is_closing():
+            return
+        self._connection_closed()
+
+    def _data_received(self, data):
+        if self._closed:
+            return
+
+        if self._reading_paused:
+            self._input_buffer.append(data)
+        elif self._protocol is not None:
+            self._protocol.data_received(data)
+
+    def _process_iq(self, payload):
+        if payload.seq != self._incoming_seq:
+            raise errors.XMPPCancelError(
+                condition=errors.ErrorCondition.UNEXPECTED_REQUEST
+            )
+        self._incoming_seq += 1
+        self._data_received(payload.content)
+
+    def _process_msg(self, payload):
+        if payload.seq != self._incoming_seq:
+            return
+        self._incoming_seq += 1
+        self._data_received(payload.content)
+
+
+class IBBService(service.Service):
+    """
+    A service implementing in-band bytestreams.
+
+    .. attribute:: SESSION_LIMIT = 0
```

Comment:
Don’t use ALL_CAPS for non-globals. The ``SESSION_LIMIT`` is an instance attribute, not a class attribute.